plugins/discovery-ec2/src/main/java/org/opensearch/discovery/ec2/AwsEc2ServiceImpl.java

```diff
@@ -32,82 +32,135 @@
+        ClientOverrideConfiguration overrideConfiguration,
+        String endpoint,
+        Region region,
+        long readTimeoutMillis
+    ) {
+        ApacheHttpClient.Builder clientBuilder = ApacheHttpClient.builder()
+            .proxyConfiguration(proxyConfiguration)
+            .socketTimeout(Duration.ofMillis(readTimeoutMillis));
+
+        Ec2ClientBuilder builder = Ec2Client.builder()
+            .overrideConfiguration(overrideConfiguration)
+            .httpClientBuilder(clientBuilder)
+            .credentialsProvider(awsCredentialsProvider)
+            .region(region);
+
         if (Strings.hasText(endpoint)) {
             logger.debug("using explicit ec2 endpoint [{}]", endpoint);
-            builder.withEndpointConfiguration(new AwsClientBuilder.EndpointConfiguration(endpoint, null));
+            builder.endpointOverride(URI.create(endpoint));
         }
+
         return SocketAccess.doPrivileged(builder::build);
     }
 
-    // pkg private for tests
-    static ClientConfiguration buildConfiguration(Logger logger, Ec2ClientSettings clientSettings) {
-        final ClientConfiguration clientConfiguration = new ClientConfiguration();
-        // the response metadata cache is only there for diagnostics purposes,
-        // but can force objects from every response to the old generation.
-        clientConfiguration.setResponseMetadataCacheSize(0);
-        clientConfiguration.setProtocol(clientSettings.protocol);
+    static ProxyConfiguration buildProxyConfiguration(Logger logger, Ec2ClientSettings clientSettings) {
         if (Strings.hasText(clientSettings.proxyHost)) {
-            // TODO: remove this leniency, these settings should exist together and be validated
-            clientConfiguration.setProxyHost(clientSettings.proxyHost);
-            clientConfiguration.setProxyPort(clientSettings.proxyPort);
-            clientConfiguration.setProxyUsername(clientSettings.proxyUsername);
-            clientConfiguration.setProxyPassword(clientSettings.proxyPassword);
+            try {
+                // TODO: remove this leniency, these settings should exist together and be validated
+                return ProxyConfiguration.builder()
+                    .endpoint(
+                        new URI(
+                            clientSettings.protocol.toString(),
+                            null,
+                            clientSettings.proxyHost,
+                            clientSettings.proxyPort,
+                            null,
+                            null,
+                            null
+                        )
+                    )
+                    .username(clientSettings.proxyUsername)
+                    .password(clientSettings.proxyPassword)
+                    .build();
+            } catch (URISyntaxException e) {
+                logger.error("Exception during URI construction for specified proxy", e);
+                throw new RuntimeException(e);
```

Comment:
Shouldn't we just throw the ex and let the consumer log?

Reply:
It would lose the fact that the problem is proxy URL. I've slightly improved the error message, changing to `Invalid proxy URL`.

Reply:
How about?

```
throw new RuntimeException("Invalid proxy URL", e);
```

The big thing is to avoid printing the stack trace to the log twice as that adds a ton of noise.